# CodonMuSel
CodonMuSel is a tool for analysing the factors affecting codon bias both at the genome-wide level and at the gene-level. It estimates 

1) Estimates the strength of selection acting on nucleotide biosynthesis cost.
2) Estimates the strength of selection acting on translational efficiency.
3) Estimates the proportion of codon bias attributable to mutational processes.
4) Determines the extent to which individual gene sequences are optimised for cost efficiency.
5) Determines the extent to which individual gene sequences are optimised for translational efficiency.
6) Determines the extent to which individual gene sequences are jointly optimised for both cost and translational efficiency.

For more details please see:

CodonMuSel implements the SK model for codon bias first described in:
Seward EA, Kelly S (2016) Dietary nitrogen alters codon bias and genome composition in parasitic microorganisms. Genome Biol 17(1):226.

AND

Seward EA, Kelly S (in prep.) The evolutionary economics of a gene, balancing the cost and efficiency of protein production.

***Running CodonMuSel:***

python CodonMuSel.py [OPTIONS] -f \<SEQUENCE FILE\> -tscan \<tRNAscan FILE\>

	-f <FILE>	A FASTA file of protein coding nucleotide seqeunces
	-tscan <FILE>	A tRNA copy number file produced by tRNAscan
	-tc <INT>	The NCBI genetic code identifier goo.gl/ByQOau (The Standard Code = 1)
	-ind		Analysed individual genes in adition to a genomewide analysis
	-fix_mb		Fix mutation bias to value obtained from genome-wide analysis when analysing individual genes
	-par 		Determine the cost and efficiency optimality of individual genes

***Input Files:***

To run a complete CodonMuSel analysis requires two input files

*1) A FASTA file containing nucleotide sequences for protein coding genes*
Typically this is the full set of protein coding genes from a single species. However, the method can be applied to sequences obtain from transcriptome or metagenome sequencing. By default CodonMuSel only analyses nucleotide sequences that begin with a start codon, end with stop codon, and do not contain in-frame stop codons. Sequences not meeting all of these criteria are ignored.

*2) tRNA copy number file generated by tRNAscan*
This is a text file containing the output generated by running tRNAscan on the complete genome of the species in question. If you do not include this optional file then CodonMuSel cannot infer the strength of selection acting on codon translational efficiency or determine the extent to which gene sequences haveo been ptimised for translational efficiency.

Schattner P, Brooks AN, Lowe TM (2005) The tRNAscan-SE, snoscan and snoGPS web servers for the detection of tRNAs and snoRNAs. Nucleic Acids Res 33:686â€“689


***Results files:***

\<SEQUENCE FILE\>_results_Bestmodelparameters.txt
\<SEQUENCE FILE\>_results_file_individual_genes_Bestmodel_parameters.txt
\<SEQUENCE FILE\>_Paerto_optimisation_results.txt


***Example Dataset:***

An example dataset is provided. To run CodonMuSel on this dataset execute the following command

python Codon_MuSel.py -f Mycoplasma_pneumoniae.fasta -tscan Mycoplasma_pneumoniae_tRNAscan.txt -tc 1 -ind -fix_mb -par
